### PR TITLE
Fix Issue 8472 - Replace walkLength() with an improved count()

### PR DESCRIPTION
**Pull request by RazvanN7**
I am not sure how the deprecation cycle works, but I deprecated the first walkLength overload which is obsolete. I don't know if the walkLength overload which walks up to a given number of elements should be deprecated too. I think that the best way is to deprecated one overload of walkLength and keep the other.

### CODE REVIEW

File: std/algorithm/searching.d [hunk 1]
```diff
@@ -696,11 +696,16 @@ size_t count(alias pred = "true", R)(R haystack)
 if (isInputRange!R && !isInfinite!R &&
     is(typeof(unaryFun!pred(haystack.front)) : bool))
 {
-    size_t result;
-    alias T = ElementType!R; //For narrow strings forces dchar iteration
-    foreach (T elem; haystack)
-        if (unaryFun!pred(elem)) ++result;
-    return result;
+    static if (hasLength!R && is(typeof(unary!fun(pred) == unary!fun("true"))))
```
Comment (wilzbach) [inline std/algorithm/searching.d]: Your `is` swallows all errors (even your typos), but comparing functions doesn't work in D, e.g.:

```
void main(string[] args)
{
    import std.functional;
    static assert(is(typeof(unaryFun!("true") == unaryFun!("true"))));
}
```

Regarding the specialization, you should see https://github.com/dlang/phobos/pull/4265#issuecomment-223762623 and its final version (https://github.com/dlang/phobos/pull/5001).
tl;dr: use overloads, not strings

File: std/range/primitives.d [hunk 2]
```diff
@@ -1638,6 +1638,7 @@ upTo) steps have been taken and returns $(D upTo).
 Infinite ranges are compatible, provided the parameter $(D upTo) is
 specified, in which case the implementation simply returns upTo.
  */
+deprecated("use std.algorithm.searching.count instead.")
```
Comment (wilzbach) [inline std/range/primitives.d]: You should use the deprecated string, so it's easy to grep for:

```
// Explicitly undocumented. It will be removed in July 2018. @@@DEPRECATED_2018-07@@@
```

And of course, undocument or state in the document with `$(RED)` that it's deprecated.

File: std/algorithm/searching.d (hunk 1 above)
Comment (RazvanN7) [inline std/algorithm/searching.d]: I guess in this case I should create an overload which doesn't receive a string function, but calls the one I just modified

**Comment by JackStouffer** [comment]
@RazvanN7 See https://issues.dlang.org/show_bug.cgi?id=11429

**Comment by RazvanN7** [comment]
I suggest keeping the remaining walkLength and deprecating the one that count can take care off

**Comment by RazvanN7** [comment]
I'm guessing the jenkins failure is due to the fact that the use of deprecated functions is forbidden in phobos?

**Comment by wilzbach** [comment]
> I'm guessing the jenkins failure is due to the fact that the use of deprecated functions is forbidden in phobos?

No it's simply broken atm (or has a lot of transient failures), however CircleCi isn't:

```
std/range/primitives.d(1656:6)[warn]: Public declaration 'walkLength' is undocumented.
```

> due to the fact that the use of deprecated functions is forbidden in phobos?

Wishful thinking -> https://github.com/dlang/phobos/pull/5546

**Comment by RazvanN7** [comment]
@wilzbach seems kind of awkward to document deprecated functions.

**Comment by jmdavis** [comment]
Please do not deprecate `walkLength`. If `count` can do what what `walkLength` does, then fine. But `walkLength` has been around since the beginning of ranges in Phobos, and it will break a _lot_ of code if folks are forced to change their code, because `walkLength` has been deprecated. This seems like a reasonable optimization to `count`, but I don't think that anything should be done to `walkLength` because of it.

File: std/range/primitives.d (hunk 2 above)
Comment (jmdavis) [inline std/range/primitives.d]: Also, that deprecated string is for symbols that have already been through a year of deprecation and are now undocumented. So, even if we deprecated `walkLength` (which IMHO, we shouldn't), this would not be the message to use. It would have to be one from a symbol that's still documented with a ddoc comment.

**Comment by wilzbach** [comment]
> @wilzbach seems kind of awkward to document deprecated functions.

-> https://github.com/dlang-community/D-Scanner/issues/492

**Comment by wilzbach** [comment]
@RazvanN7 I suggest to split this PR and submit the optimization for `count` in another as that one isn't controversial and can be merged quickly.

**Comment by jmdavis** [comment]
Also, if `walkLength` isn't being deprecated, it would be better for `count` to simply call `walkLength` in the new overload rather than reimplementing it, because there's less code duplication that way.

**Comment by andralex** [comment]
@RazvanN7 deprecation of anything is a tall order and should be confined to its own pull request (as opposed to mixed with other work). I'll close this and comment in the other.
